### PR TITLE
Feature + Civilian Research Tree

### LINKED ISSUE #4504: Pathing into oil
### Minecolonies version
This is on the Patreon 1.15 server.
Minecolonies: 
minecolonies-0.11.787-ALPHA-universal.jar

Before an update about a week ago, my deliveryman used to walk this way all the time without problems. But now when he needs to walk in this direction, he just walks right into an oil pit, bounces a few times, then sits there. I believe the oil pits are created by Industrial Foregoing? Unsure if he'd have the same problem in water.

He's very enthusiastic about diving into the oil... lol

### LINKED ISSUE #4588: Non-stackable Research cost
### Minecolonies version

Version: Newest

### Expected behavior

- 

### Actual behaviour

- 
Some researches like Archer mastery 64 bows require nonstackable items, not sure if its only the archer mastery

### Steps to reproduce the problem

1. 
2. 
3. 
4. ... Profit

### Logs: Latest.log and crashlog at https://gist.github.com/

- latest.log:
- crashlog:

### LINKED ISSUE #4596: [1.15] Medieval Warehouse and Deliveryhut conflict
Version: minecolonies-0.11.817-ALPHA-universal

### Expected behavior
Deliveryman takes items from other huts (or the new stash which is very nice btw) and stores them in the warehouse.

### Actual behaviour
The Deliveryman takes items, goes to the warehouse but they are placed in the chest which is part od the delivery house. Then he goes back to his house, takes the items again - goes to the warehouse etc...

After many tries - in this case 4 or 5 repeats the items end somewhere else (in the warehouse hopefully) and not back in the delivery hut.

I often see the racks and chests in this delivery hut full of random items (collected mob drop, excess building materials returned from the builder, resources gathered by other huts etc).

### Steps to reproduce the problem
1. Build a medieval (Oak) warehouse (level 1)
2. Build a medieval (oak) deliveryman hut on the indicated spot (which overlaps with the warehouse).
3. not sure if required - but I upgraded my deliveryman to level 3 while the warehouse is still at 1.
4. ... Rename your delivery guy to Sisyphus

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #4504
Closes #4588
Closes #4596

The second part of the research tree

# Changes proposed in this pull request:
- Civilian Research Path
- Add Rails Pathing

Review please

### CODE REVIEW

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java  11
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  1
- src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java  39
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  1
- src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5334784&bid=17462552)

File: src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+import net.minecraftforge.fml.network.NetworkHooks;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+/**
+ * Special minecolonies minecart that doesn't collide.
+ */
+public class MinecoloniesMinecart extends AbstractMinecartEntity
```
Comment (Nightenom) [inline src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java:28]: Why do we need that? Can't it extend Minecart entity instead? (looks way much copy code to me)

File: src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+{
+    /**
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java:30]: Do we want this to ignore powered rails and such? should overwrite 
    /**
     * Return false if this cart should not call onMinecartPass() and should ignore Powered Rails.
     * @return True if this cart should call onMinecartPass().
     */
    default boolean shouldDoRailFunctions() {
        return true;
    }
then

File: src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java (hunk 1 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java:28]: No, we need to alter a bunch of code in-between unfortunately to make it work nicely for our usecase. So we already usecase as much as we could.

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java [hunk 3]
```diff
@@ -410,6 +450,105 @@ else if (ourEntity.isInWater())
...
+
+            if (pEx.isOnRails() || pEx.isRailsExit())
+            {
+                return handlePathOnRails(pEx, pExNext);
+            }
+        }
+        return false;
+    }
+
+    private boolean handlePathOnRails(final PathPointExtended pEx, final PathPointExtended pExNext)
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java:483]: javadoc

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java [hunk 4]
```diff
@@ -532,18 +671,50 @@ protected void pathFollow()
...
     @Override
-    protected void checkForStuck(Vec3d positionVec3)
+    protected void checkForStuck(@NotNull final Vec3d positionVec3)
     {
         if (world.getGameTime() - pathStartTime < MIN_KEEP_TIME)
         {
             return;
         }
 
-        super.checkForStuck(positionVec3);
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java:546]: think clearpath could use some dismount when it somehow gets stuck within the minecart(make them timed lifetime aswell?)

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java [hunk 5]
```diff
@@ -410,6 +450,105 @@ else if (ourEntity.isInWater())
...
+                    break;
+                case NORTH:
+                    forward = Math.max(Math.min(motion.getZ() + 0.01D, 1), 0);
+                    entity.ridingEntity.setMotion(motion.add(0.0D, 0.0D, forward == 1 ? 1 : 0.01D));
+                    break;
+                case SOUTH:
+                    forward = Math.min(Math.max(motion.getZ() - 1 * 0.01D, -1), 0);
+                    entity.ridingEntity.setMotion(motion.add(0.0D, 0.0D, forward == -1 ? -1 : -1 * 0.01D));
+            }
+            //todo add research branch civilian where we add using railcarts as an option to guard this.
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java]: old todo?

File: src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java (hunk 2 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java:30]: Wanted to keep it as is for now. If we notice problems with them we can still turn them off.

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java (hunk 4 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java:546]: they will automatically dismount when they try a new path that doesn't take the rail.

File: src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java]: yes

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  1
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  1
- src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java  4
- src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java  11
- src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java  39
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5334784&bid=17462552)

**Review by someaddons — DISMISSED** [review]

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java  4
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  1
- src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java  11
- src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java  39
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5334784&bid=17462552)

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  1
- src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java  11
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  1
- src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java  39
- src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5334784&bid=17462552)

**Comment by marvin-bitterlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/pathfinding/MinecoloniesAdvancedPathNavigate.java  11
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  1
- src/main/java/com/minecolonies/coremod/entity/pathfinding/AbstractPathJob.java  4
- src/api/java/com/minecolonies/api/entity/MinecoloniesMinecart.java  39
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5334784&bid=17462552)

**Review by someaddons — APPROVED** [review]

**Pull request merged**